Source: based on Gentoo's ocaml-3.10.0-exec-stack-fixes.patch
Upstream: no
Reason: set progbits for all supported architectures

Index: ocaml-3.11.1/asmcomp/alpha/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/alpha/emit.mlp
+++ ocaml-3.11.1/asmcomp/alpha/emit.mlp
@@ -811,6 +811,11 @@
   `	.data\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux" then
+    `  .section .note.GNU-stack,\"\",%progbits\n`
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -843,6 +848,7 @@
   `{emit_symbol lbl_begin}:\n`
 
 let end_assembly () =
+  nx_stack();
   let lbl_end = Compilenv.make_symbol (Some "code_end") in
   `	.text\n`;
   `	.globl	{emit_symbol lbl_end}\n`;
Index: ocaml-3.11.1/asmrun/alpha.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/alpha.S
+++ ocaml-3.11.1/asmrun/alpha.S
@@ -438,3 +438,7 @@
         .word   -1              /* negative frame size => use callback link */
         .word   0               /* no roots here */
         .align  3
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
Index: ocaml-3.11.1/asmrun/arm.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/arm.S
+++ ocaml-3.11.1/asmrun/arm.S
@@ -342,3 +342,7 @@
         .short  -1              /* negative frame size => use callback link */
         .short  0               /* no roots */
         .align  2
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
Index: ocaml-3.11.1/asmrun/hppa.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/hppa.S
+++ ocaml-3.11.1/asmrun/hppa.S
@@ -532,3 +532,8 @@
         .long   L104 + 3        /* return address into callback */
         .short  -1              /* negative frame size => use callback link */
         .short  0               /* no roots */
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
+
Index: ocaml-3.11.1/asmrun/ia64.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/ia64.S
+++ ocaml-3.11.1/asmrun/ia64.S
@@ -528,3 +528,8 @@
 
         .common caml_saved_bsp#, 8, 8
         .common caml_saved_rnat#, 8, 8
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
+
Index: ocaml-3.11.1/asmrun/m68k.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/m68k.S
+++ ocaml-3.11.1/asmrun/m68k.S
@@ -242,3 +242,8 @@
         .long   L107            | return address into callback
         .word   -1              | negative frame size => use callback link
         .word   0               | no roots here
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
+
Index: ocaml-3.11.1/asmrun/power-elf.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/power-elf.S
+++ ocaml-3.11.1/asmrun/power-elf.S
@@ -419,3 +419,7 @@
         .short  -1              /* negative size count => use callback link */
         .short  0               /* no roots here */
 
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
+
Index: ocaml-3.11.1/asmrun/sparc.S
===================================================================
--- ocaml-3.11.1.orig/asmrun/sparc.S
+++ ocaml-3.11.1/asmrun/sparc.S
@@ -405,3 +405,8 @@
         .type Caml_raise_exception, #function
 	.type Caml_system__frametable, #object
 #endif
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
+
Index: ocaml-3.11.1/asmcomp/amd64/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/amd64/emit.mlp
+++ ocaml-3.11.1/asmcomp/amd64/emit.mlp
@@ -731,6 +731,13 @@
   `	.data\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux" then
+    `  .section .note.GNU-stack,\"\",%progbits\n`
+
+
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -761,6 +768,7 @@
   if macosx then `	nop\n` (* PR#4690 *)
 
 let end_assembly() =
+  nx_stack();
   let lbl_end = Compilenv.make_symbol (Some "code_end") in
   `	.text\n`;
   if macosx then `	nop\n`; (* suppress "ld warning: atom sorting error" *)
Index: ocaml-3.11.1/asmcomp/arm/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/arm/emit.mlp
+++ ocaml-3.11.1/asmcomp/arm/emit.mlp
@@ -642,6 +642,13 @@
   `	.data\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux" then
+    `  .section .note.GNU-stack,\"\",%progbits\n`
+
+
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -658,6 +665,7 @@
   `{emit_symbol lbl_begin}:\n`
 
 let end_assembly () =
+  nx_stack();
   let lbl_end = Compilenv.make_symbol (Some "code_end") in
   `	.text\n`;
   `	.global	{emit_symbol lbl_end}\n`;
Index: ocaml-3.11.1/asmcomp/hppa/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/hppa/emit.mlp
+++ ocaml-3.11.1/asmcomp/hppa/emit.mlp
@@ -994,6 +994,12 @@
   `	.data\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux" then
+      `  .section .note.GNU-stack,\"\",%progbits\n`
+
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -1022,6 +1028,7 @@
 
 
 let end_assembly() =
+  nx_stack();
   `	.code\n`;
   let lbl_end = Compilenv.make_symbol (Some "code_end") in
   declare_global lbl_end;
Index: ocaml-3.11.1/asmcomp/i386/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/i386/emit.mlp
+++ ocaml-3.11.1/asmcomp/i386/emit.mlp
@@ -948,6 +948,11 @@
   `	.data\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux_elf" then
+      `  .section .note.GNU-stack,\"\",%progbits\n`
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -962,6 +967,7 @@
   if macosx then `	nop\n` (* PR#4690 *)
 
 let end_assembly() =
+  nx_stack();
   let lbl_end = Compilenv.make_symbol (Some "code_end") in
   `	.text\n`;
   if macosx then `	nop\n`; (* suppress "ld warning: atom sorting error" *)
Index: ocaml-3.11.1/asmcomp/ia64/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/ia64/emit.mlp
+++ ocaml-3.11.1/asmcomp/ia64/emit.mlp
@@ -1306,6 +1306,11 @@
   `	.align 8\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux" then
+      `  .section .note.GNU-stack,\"\",%progbits\n`
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -1315,6 +1320,7 @@
   emit_define_symbol (Compilenv.make_symbol (Some "code_begin"))
 
 let end_assembly () =
+  nx_stack();
   `	.data\n`;
   emit_define_symbol (Compilenv.make_symbol (Some "data_end"));
   `	.text\n`;
Index: ocaml-3.11.1/asmcomp/power/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/power/emit.mlp
+++ ocaml-3.11.1/asmcomp/power/emit.mlp
@@ -920,6 +920,12 @@
   emit_string data_space;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "elf" then
+      `  .section .note.GNU-stack,\"\",%progbits\n`
+
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -936,6 +942,7 @@
   `{emit_symbol lbl_begin}:\n`
 
 let end_assembly() =
+  nx_stack();
   if pic_externals then
     (* Emit the pointers to external functions *)
     StringSet.iter emit_external !external_functions;
Index: ocaml-3.11.1/asmcomp/sparc/emit.mlp
===================================================================
--- ocaml-3.11.1.orig/asmcomp/sparc/emit.mlp
+++ ocaml-3.11.1/asmcomp/sparc/emit.mlp
@@ -741,6 +741,12 @@
   `	.data\n`;
   List.iter emit_item l
 
+(* Mark stack as non executable *)
+let nx_stack() =
+  if Config.system = "linux" then
+      `  .section .note.GNU-stack,\"\",%progbits\n`
+
+
 (* Beginning / end of an assembly file *)
 
 let begin_assembly() =
@@ -754,6 +760,7 @@
   `{emit_symbol lbl_begin}:\n`
 
 let end_assembly() =
+  nx_stack();
   `	.text\n`;
   let lbl_end = Compilenv.make_symbol (Some "code_end") in
   `	.global	{emit_symbol lbl_end}\n`;
